### ISSUE TITLE
SP5: The program becomes slow when drawing a million Polygons.

### ISSUE THREAD

**Issue by ch210** [ScottPlot/ScottPlot#3266]
**Question:**  I've changed V4 to V5, but the running speed of drawing millions of Polygons didn't became faster 

**ScottPlot Version:**  ScottPlot.WinForms 5.0.18 

I'm using .NET 6.0
When I just draw a 100*100 map, the running time is tiny, but if I change it to 1000*1000, the program is very slow and it's hard to zoom in, zoom out and drag the map.

100*100 Situation: 
![image](https://github.com/ScottPlot/ScottPlot/assets/143468772/83ae8035-b539-4d18-886e-f2e05cfb5479)

1000*1000 Situation: 
![image](https://github.com/ScottPlot/ScottPlot/assets/143468772/8b2ffed8-e363-43be-aa83-1b36a6224e95)

**Code Sample:** Here's the code.

```cs
using ScottPlot.Plottables;
using ScottPlot.WinForms;

namespace WinFormsApp1
{
    public partial class Form1 : Form
    {

        public FormsPlot MainPlot;
        public Form1()
        {
            InitializeComponent();
            MainPlot = new FormsPlot() { Dock = DockStyle.Fill };
            this.splitContainer1.Panel2.Controls.Add(MainPlot);
        }


        public class PanelModel
        {
            public int Row { get; set; }

            public int Column { get; set; }

            public int PanelSize { get; set; }

            public double Die_Width { get; set; }

            public double Die_Height { get; set; }
        }

        public class Die_In_Panel
        {
            public bool IsPass { get; set; }

            public string NGCode { get; set; }

            public Polygon Polygon { get; set; }

            public Dictionary<string, double> ParameterNameValueMap { get; set; }
        }

        public class SPI
        {
            public bool Is_Positive_electrode { get; set; }

            public Dictionary<string, double> ParameterNameValueMap { get; set; }
        }

        private void BtnDraw_Click(object sender, EventArgs e)
        {
            PanelModel panel = new PanelModel();
            panel.PanelSize = Convert.ToInt32(textBox1.Text) * 1000;
            panel.Row = Convert.ToInt32(textBox2.Text);
            panel.Column = Convert.ToInt32(textBox3.Text);

            System.Diagnostics.Stopwatch stopwatch = new System.Diagnostics.Stopwatch();
            stopwatch.Start();
            DrawPanelMap_ScottV5(panel);
            Console.WriteLine(stopwatch.ElapsedMilliseconds);
            stopwatch.Stop();
        }

        public ScottPlot.Color GetScottColorByArgb(Color color)
        {
            return new ScottPlot.Color(color.R, color.G, color.B,color.A);
        }
        private void DrawPanelMap_ScottV5(PanelModel panel)
        {
            MainPlot.Plot.Clear();
            MainPlot.Plot.Axes.Bottom.Min = 0;
            MainPlot.Plot.Axes.Bottom.Max = panel.PanelSize;

            MainPlot.Plot.Axes.Left.Min = 0;
            MainPlot.Plot.Axes.Left.Max = panel.PanelSize;

            ScottPlot.Color NGDieColor = GetScottColorByArgb(Color.Yellow);
            ScottPlot.Color PassDieColor = GetScottColorByArgb(Color.Green);
            ScottPlot.Color negative_Electrode_Color = GetScottColorByArgb(Color.Black);
            ScottPlot.Color positive_Electrode_Color = GetScottColorByArgb(Color.White);
            ScottPlot.Color panelFillColor = GetScottColorByArgb(Color.Transparent);
            ScottPlot.Color lineColor = GetScottColorByArgb(Color.FromArgb(50, Color.Black));

            Polygon polygon = MainPlot.Plot.Add.Polygon(new ScottPlot.Coordinates[] { new ScottPlot.Coordinates(0, 0),
            new ScottPlot.Coordinates(0, panel.PanelSize),new ScottPlot.Coordinates(panel.PanelSize, panel.PanelSize),
                new ScottPlot.Coordinates(panel.PanelSize, 0)});
            polygon.LineStyle.Width = 2;
            polygon.LineStyle.Color = negative_Electrode_Color;
            polygon.FillStyle.Color = panelFillColor;

            panel.Die_Width = panel.PanelSize / (double)panel.Column;
            panel.Die_Height = panel.PanelSize / (double)panel.Row;


            Die_In_Panel[,] dice = new Die_In_Panel[panel.Row, panel.Column];
            for (int r = 0; r < dice.GetLength(0); r++)
            {
                for (int c = 0; c < dice.GetLength(1); c++)
                {
                    dice[r, c] = new Die_In_Panel();
                    dice[r, c].IsPass = true;
                }
            }
            int romCount = 10;
            Random random = new Random();

            for (int i = 0; i < romCount; i++)
            {
                int yindex = random.Next(panel.Row);
                int xindex = random.Next(panel.Column);

                dice[yindex, xindex].IsPass = false;
            }
            int x_index = 0;
            for (double x = 0; x < panel.PanelSize; x += panel.Die_Width)
            {
                int y_index = 0;
                for (double y = 0; y < panel.PanelSize; y += panel.Die_Height)
                {
                    if (dice[y_index, x_index].IsPass)
                    {
                        dice[y_index, x_index].Polygon = MainPlot.Plot.Add.Polygon(new ScottPlot.Coordinates[]
                            {
                                new ScottPlot.Coordinates(x,y),
                                new ScottPlot.Coordinates(x,y+ panel.Die_Height),
                                new ScottPlot.Coordinates(x + panel.Die_Width,y+ panel.Die_Height),
                                new ScottPlot.Coordinates(x + panel.Die_Width,y)
                            });
                        dice[y_index, x_index].Polygon.LineStyle.Width = 1;
                        dice[y_index, x_index].Polygon.LineStyle.Color = lineColor;
                        dice[y_index, x_index].Polygon.FillStyle.Color = PassDieColor;
                    }
                    else
                    {
                        dice[y_index, x_index].Polygon = MainPlot.Plot.Add.Polygon(new ScottPlot.Coordinates[]
                            {
                                new ScottPlot.Coordinates(x,y),
                                new ScottPlot.Coordinates(x,y+ panel.Die_Height),
                                new ScottPlot.Coordinates(x + panel.Die_Width,y+ panel.Die_Height),
                                new ScottPlot.Coordinates(x + panel.Die_Width,y)
                            });
                        dice[y_index, x_index].Polygon.LineStyle.Width = 1;
                        dice[y_index, x_index].Polygon.LineStyle.Color = lineColor;
                        dice[y_index, x_index].Polygon.FillStyle.Color = NGDieColor;
                    }
                    
                    y_index++;
                }
                x_index++;
            }
            MainPlot.Refresh();
        }
    }
}

```

**Comment by swharden** [comment]
Hi @ch210, this is a pretty complicated issue and may take me a little while to get to. I probably need to resolve #3190 first as well.

In the mean time, are polygons really required? Can a heatmap (that has transparent squares) meet your use case? I suspect that is a must faster option if your data points can fit inside a rectangular grid.

**Comment by ch210** [comment]
Thanks for your reply. I'll try to finish my task with heat map. But let me describe the case that I met.
The map I draw shows the information of a large panel with millions of tiny LEDs, I'd like to see specific information about each unit with details like Voltage, Defect, Illumination, etc. Furthermore,  in erery little unit, I'll draw a couple of points (like the positive and negative terminals of a battery) that much smaller than the unit, like this: 
![image](https://github.com/ScottPlot/ScottPlot/assets/143468772/337b1a7c-5ff8-4c91-8479-2504a7dea2df)

The real difficulty is those points, they have their own information either, so I'm wondering if this could be done by a heat map.
By the way, I learned that ScottPlotV5 uses SkiaSharp with OpenGL, does this change be used when I draw so many polygons? I used both SP4 and SP5 to do the same thing, but the performance has no significant difference.
